unsuccessful whisper no feedback fix

SomeName42 opened the pull request:
Resolves #9049. The source of the error was trying to add message to the current whisper room with the add_chat_room_message_received function which is not meant to be used with whisper rooms. The error was solved by checking if the current room is whisper room and in this case selecting the lobby. There was an other error because during staging there is no lobby. This was solved by checking if the lobby room exists and if it doesnt then sending the message to the 'this game' room. Furthermore when receiving the cant find user message from the server the associated whisper room is closed.

Commit:
Lobby and game creation chatbox

Fixed no feedback when whispering to non existing user. The source of the error was trying to add message to the current whisper room with the add_chat_room_message_received function which is not meant to be used with whisper rooms. The error was solved by checking if the current room is whisper room and in this case selecting the lobby. There was an other error because during staging there is no lobby. This was solved by checking if the lobby room exists and if it doesnt then sending the message to the 'this game' room. Furthermore when receiving the cant find user message from the server the associated whiser room is closed.

SomeName42 commented:
Would it be better if all messages without rooms would always be added to the lobby or the this game rooms, and not the current?

soliton- commented:
Does it make a difference? What is the current?

Btw, there is currently no room support anymore on the server side.

Server error messages should always go to the lobby or 'this game' (pre-game lobby) I suppose.

Commit:
Lobby and game creation chatbox

Removed unnecessary code for removed room support. Fixed whisper room with lobby or this game player name unable to be closed.

soliton- commented:
Please format commit messages like this:

```
<subject>

<rest of commit message with more details>
```

https://wiki.wesnoth.org/DeveloperGuide#Commit_messages

soliton- commented on src/gui/widgets/chatbox.cpp:
```diff
@@ -587,20 +587,18 @@ void chatbox::process_message(const ::config& data, bool whisper /*= false*/)
-		if(room.empty()) {
-			LOG_LB << "Message without a room from " << sender << ", falling back to active window";
-			room = open_windows_[active_window_].name;
+			close_window_button_callback(message.substr(12, message.size() - 14), dummy_var, dummy_var);
```
What is this for?

SomeName42 replied:
It closes the whisper room using the close button callback.

SomeName42 replied:
When the server sends the cant find error in response to the whisper message.

soliton- replied:
That seems like a confusing thing to do. Now you cannot see the message you sent at all anymore? What if you want to resend it once the other client is online again? (Unfortunately we don't have history functionality here either for some reason.)

SomeName42 replied:
Then would it be better If the can't find player error message be displayed in the whisper room? If it is displayed in the lobby or this game room it could be easy to miss when the whisper room is active.

soliton- replied:
I don't like the message parsing but that would be useful. I'd rather extend the message format and add a `original_receiver` attribute to the error message that the client can use to find a fitting tab to display the message.

Commit:
wesnothd, lobby and game creation chatbox

Made server set original_receiver field when sending whisper error messages.
Made chatbox use original_receiver field to display messages in whisper rooms.
Removed whisper room closing on cant find player error message.

SomeName42 commented:
I think actually it would be better to filter messages based on sender and type and display some in the lobby or this game rooms and others in the current room as there are other commands that the server sends replies to and because when response in created on client side it also appears in the current room.

soliton- commented:
If this means that messages appear in random tabs based on what was selected at the time of arrival then that does not make any sense to me.

soliton- commented on src/gui/widgets/chatbox.cpp:
```diff
@@ -587,23 +596,27 @@ void chatbox::process_message(const ::config& data, bool whisper /*= false*/)
-		// If we still don't have a name, fall back to lobby.
-		if(room.empty()) {
-			LOG_LB << "Message without a room from " << sender << ", assuming lobby";
+		if((sender == "server" && type == "motd") || sender != "server") {
```
Server messages do not go into the lobby unless it's the motd? What is this logic based on?

SomeName42 replied:
This is the implementation for the idea described in my comment. It tries to separate responses to commands that the user runs so that they are displayed in the current room and are visible.

SomeName42 commented:
> If this means that messages appear in random tabs based on what was selected at the time of arrival then that does not make any sense to me.

The idea is the same as with the whisper error message. When the user runs a command if they don't have the lobby/"this game" room selected they might miss the reply.

soliton- commented:
> The idea is the same as with the whisper error message. When the user runs a command if they don't have the lobby/"this game" room selected they might miss the reply.

The idea from the bug report? The bug report is about not showing the error message at all. Showing the error in the lobby (or "this game") is ok. Showing it in the tab with the message that triggered the error is even better.

> When commands are handled on the client side it also appears to work the same way.

Can you give an example?

SomeName42 commented:
> > The idea is the same as with the whisper error message. When the user runs a command if they don't have the lobby/"this game" room selected they might miss the reply.
> 
> The idea from the bug report? The bug report is about not showing the error message at all. Showing the error in the lobby (or "this game") is ok. Showing it in the tab with the message that triggered the error is even better.
> 
> > When commands are handled on the client side it also appears to work the same way.
> 
> Can you give an example?

For example the /help command or when you give a command wrong number of arguments.
Would it be an acceptable solution then if it was handled with a new field in the server message such as "command_response"

soliton- commented:
> For example the /help command or when you give a command wrong number of arguments. Would it be an acceptable solution then if it was handled with a new field in the server message such as "command_response"?

I'm not sure how this is related? A local command typed into some tab can certainly also output whatever in that tab. This has nothing to do with a server message at all.

SomeName42 commented:
> > For example the /help command or when you give a command wrong number of arguments. Would it be an acceptable solution then if it was handled with a new field in the server message such as "command_response"?
> 
> I'm not sure how this is related? A local command typed into some tab can certainly also output whatever in that tab. This has nothing to do with a server message at all.

From the users perspective I don't think it makes a difference whether the command was processed locally or on the server. All the user sees is the output. I also think that if the command output appears in a different room than is currently open when the command is issued it could easily be missed especially if in other cases it does appear in the current room. My idea is to separate server command responses and only make those appear in the current room. I agree in the previous commit it was not done properly, but I think in the current one it is or is done much better.

soliton- commented:
> From the users perspective I don't think it makes a difference whether the command was processed locally or on the server. All the user sees is the output. I also think that if the command output appears in a different room than is currently open when the command is issued it could easily be missed especially if in other cases it does appear in the current room.

How do you get `/help` output in a different window? There is a principal difference between local commands that are immediately processed and commands sent to the server that get an independent response that might also take a bit of time to arrive.

> My idea is to separate server command responses and only make those appear in the current room.

I don't think that is sensible. Either we show server command responses in the window the command was issued as you yourself hinted at above (or if the command created a window in that window) or we always show them in the lobby.

I would also be happy with just fixing the bug from #9049.

SomeName42 commented:
Reverted to the solution with the added original receiver field

soliton- commented:
Looks generally good. I wouldn't want to squash everything to one commit though. It would be nice if you could rewrite this to fewer logical commits. If not I can try to do that at some point.

I'm also not sure what the `add_chat_room_message_received_general` thing is about or whether that is still necessary.

SomeName42 commented:
> Looks generally good. I wouldn't want to squash everything to one commit though. It would be nice if you could rewrite this to fewer logical commits. If not I can try to do that at some point.
> 
> I'm also not sure what the `add_chat_room_message_received_general` thing is about or whether that is still necessary.

The add_chat_room_message_received_general is needed to be able to add the whisper error messages to the whisper rooms. For this it needs an additional parameter. The add_chat_room_message_received originally only worked with non whisper rooms and I left it that way to avoid changing the function in parent class as the additional parameter isnt needed there.

Pentarctagon commented:
Anything left to do before merging this?

SomeName42 commented:
> Anything left to do before merging this?

I think it should be all good.

Pentarctagon commented:
@soliton- any issues?
